### PR TITLE
metric function changes

### PR DESCRIPTION
**Pull request by UpadhyayaAnkur**

### CODE REVIEW

File: metrics/kafka.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+func PublishMsgEvent(msgType mi.SmfMsgType) error {
+
+	if nfInstanceId == "" {
+		nfInstanceId = os.Getenv("SMF_ID")
+	}
+
+	smKafkaMsgEvt := mi.MetricEvent{EventType: mi.CMsgTypeEvt, MsgType: mi.CoreMsgType{MsgType: msgType.String(), SourceNfId: nfInstanceId}}
+	if msg, err := json.Marshal(smKafkaMsgEvt); err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (UpadhyayaAnkur) [reply metrics/kafka.go:93]: Done

File: metrics/kafka.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+func (writer Writer) PublishPduSessEvent(ctxt mi.CoreSubscriber, op mi.SubscriberOp) error {
+
+	smKafkaEvt := mi.MetricEvent{EventType: mi.CSubscriberEvt,
+		SubscriberData: mi.CoreSubscriberData{Subscriber: ctxt, Operation: op}}
+	if msg, err := json.Marshal(smKafkaEvt); err != nil {
+		logger.KafkaLog.Errorf("publishing pdu sess event error [%v] ", err.Error())
+		return err
+	} else {
+		logger.KafkaLog.Debugf("publishing pdu sess event[%s] ", msg)
+		StatWriter.SendMessage(msg)
```
[the remark this answers is not in the document]
Comment (UpadhyayaAnkur) [reply metrics/kafka.go:78]: Added write error log in SendMessage()

File: metrics/kafka.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+		logger.KafkaLog.Debugf("publishing msg event[%s] ", msg)
+		StatWriter.SendMessage(msg)
+	}
+	return nil
+}
+
+func (writer Writer) PublishNfStatusEvent(msgEvent mi.MetricEvent) error {
+
+	if msg, err := json.Marshal(msgEvent); err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (UpadhyayaAnkur) [reply metrics/kafka.go:104]: Added log

File: context/context.go [hunk 4]
```diff
@@ -26,6 +26,7 @@ import (
 
 func init() {
 	smfContext.NfInstanceID = uuid.New().String()
+	os.Setenv("SMF_ID", smfContext.NfInstanceID)
```
Comment (thakurajayL) [inline context/context.go:29]: Would be interested in knowing if this is good practice to set & use environment variables.

File: context/sm_context.go [hunk 5]
```diff
@@ -681,3 +686,70 @@ func (smContext *SMContext) CommitSmPolicyDecision(status bool) error {
...
+
+	//Populate kafka sm ctxt struct
+	kafkaSmCtxt.Imsi = smContext.Supi
+	kafkaSmCtxt.IPAddress = smContext.PDUAddress.Ip.String()
+	kafkaSmCtxt.SmfSubState, op = mapPduSessStateToMetricStateAndOp(smContext.SMContextState)
+	kafkaSmCtxt.SmfId = smContext.Ref
+	kafkaSmCtxt.Slice = "sd:" + smContext.Snssai.Sd + " sst:" + strconv.Itoa(int(smContext.Snssai.Sst))
+	kafkaSmCtxt.Dnn = smContext.Dnn
+	kafkaSmCtxt.UpfName, kafkaSmCtxt.UpfAddr = smContext.getSmCtxtUpf()
+	//kafkaSmCtxt.SmfId = SMF_Self().NfInstanceID
```
Comment (thakurajayL) [inline context/sm_context.go:722]: Do you want to get rid of commented code ?

File: context/sm_context.go [hunk 6]
```diff
@@ -681,3 +686,70 @@ func (smContext *SMContext) CommitSmPolicyDecision(status bool) error {
...
+	kafkaSmCtxt.SmfIp = os.Getenv("POD_IP")
```
Comment (thakurajayL) [inline context/sm_context.go:723]: Wish to know how costly it is to read env variable like this vs having this Information available at context level.

File: metrics/kafka.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+var StatWriter Writer
+
+func InitialiseKafkaStream(config *factory.Configuration) error {
+
+	brokerUrl := "sd-core-kafka-headless:9092"
+	topicName := "sdcore-data-source-smf"
+
+	if config.KafkaInfo.BrokerUri != "" && config.KafkaInfo.BrokerPort != 0 {
+		brokerUrl = fmt.Sprintf("%s:%d", config.KafkaInfo.BrokerUri, config.KafkaInfo.BrokerPort)
+		logger.KafkaLog.Debugf("initialise kafka broker url [%v]", brokerUrl)
```
Comment (thakurajayL) [inline metrics/kafka.go:36]: Log outside if loop will help. Since you have default value for topic & URL.

**Review by thakurajayL — CHANGES_REQUESTED** [review]

File: context/sm_context.go (hunk 5 above)
Comment (UpadhyayaAnkur) [reply context/sm_context.go:722]: Yes, done

File: context/sm_context.go (hunk 6 above)
Comment (UpadhyayaAnkur) [reply context/sm_context.go:723]: This shall be removed.

File: upfadapter/pfcp/udp/udp.go [hunk 8]
```diff
@@ -23,21 +23,6 @@ var ServerStartTime time.Time
 var CPNodeID *pfcpType.NodeID
 
 func init() {
-
-	/*podIpStr := os.Getenv("POD_IP")
```
Comment (UpadhyayaAnkur) [inline upfadapter/pfcp/udp/udp.go:27]: removed commented code

**Review by thakurajayL — APPROVED** [review]

**Pull request merged**